### PR TITLE
[feat] Import examples from aws-samples/aws-lambda-powertools-examples

### PR DESCRIPTION
**Pull request by scottgerring**
**Issue #, if available:** 
#1050

## Description of changes:
Moving the samples from https://github.com/aws-samples/aws-lambda-powertools-examples into the repository. This will help make the samples more discoverable, allow us to keep them up-to-date with the powertools themselves, and follow the pattern established by the python powertools.

**Checklist**

 

* [x] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-java/#tenets)
* [x] Update tests
* [x] Update docs
* [x] PR title follows [conventional commit semantics](https://www.conventionalcommits.org/en/v1.0.0/)

## Detailed Description

The code is lifted verbatim from the old repository. I've tied it all together into one big maven build tree, and introduced an intermediate pom in the examples project. This has included some effort to restructure the examples themselves so they match the general maven project structure. 

## Breaking change checklist

**RFC issue #**:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -80,8 +80,8 @@ And configure the aspectj-maven-plugin to compile-time weave (CTW) the aws-lambd
 
 ## Example
 
-See **[example](https://github.com/aws-samples/aws-lambda-powertools-examples/tree/main/java)** for example project showcasing usage of different utilities. 
-Have a demo project to contribute which showcase usage of different utilities from powertools? We are happy to accept it [here](https://github.com/aws-samples/aws-lambda-powertools-examples/blob/main/CONTRIBUTING.md#security-issue-notifications).
+See the **[examples](examples)** directory example project showcasing usage of different utilities.
```
Comment (scottgerring) [inline README.md:83]: Point readme back to the nested samples rather than the old repo

File: examples/pom.xml [hunk 2]
```diff
@@ -0,0 +1,53 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (scottgerring) [inline examples/pom.xml:1]: Introducing an intermediate parent pom to aggregate the samples together

File: examples/powertools-examples-core/README.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
+# CoreUtilities
```
Comment (scottgerring) [inline examples/powertools-examples-core/README.md:1]: These and the reste of the concrete samples have simply been lifted from `main` from the linked samples repository.

https://github.com/aws-samples/aws-lambda-powertools-examples

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -42,7 +47,7 @@ jobs:
       max-parallel: 4
       matrix:
         # test against latest update of each major Java version, as well as specific updates of LTS versions:
-        java: [8, 8.0.192, 11.0.x, 11.0.3, 12, 13, 15, 16, 17 ]
```
Comment (scottgerring) [inline .github/workflows/build.yml:45]: Having trouble with java 8 with the examples build. I will put this back in once i've fixed the issue.

File: spotbugs-exclude.xml [hunk 5]
```diff
@@ -152,4 +152,13 @@
         <Class name="software.amazon.lambda.powertools.logging.internal.LambdaLoggingAspect"/>
         <Method name="setLogLevelBasedOnSamplingRate"/>
     </Match>
+    <Match>
```
Comment (scottgerring) [inline spotbugs-exclude.xml:155]: Spotbugs wasn't run on the examples previously. This lint comes from the original projects and I have not yet corrected it.

File: examples/powertools-examples-sqs/src/main/java/org/demo/sqs/SqsMessageSender.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+public class SqsMessageSender implements RequestHandler<ScheduledEvent, String> {
+
+    private static final Logger log = LogManager.getLogger(SqsMessageSender.class);
+
+    private static final SqsClient sqsClient = SqsClient.builder()
+            .httpClient(UrlConnectionHttpClient.create())
+            .build();
+
+    private static final Random random = new Random();
```
Comment (stevehouel) [inline examples/powertools-examples-sqs/src/main/java/org/demo/sqs/SqsMessageSender.java:36]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

To make your code secure, use `new SecureRandom()` to instantiate a `java.security.SecureRandom` object. Do not use `getInstance()`, `getInstanceStrong()` and `java.util.Random`.

File: examples/powertools-examples-sqs/src/main/java/org/demo/sqs/SqsPoller.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+import static com.amazonaws.services.lambda.runtime.events.SQSEvent.SQSMessage;
+
+/**
+ * Handler for requests to Lambda function.
+ */
+public class SqsPoller implements RequestHandler<SQSEvent, String> {
+
+    Logger log = LogManager.getLogger(SqsPoller.class);
+    Random random = new Random(100);
```
Comment (stevehouel) [inline examples/powertools-examples-sqs/src/main/java/org/demo/sqs/SqsPoller.java:25]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

To make your code secure, use `new SecureRandom()` to instantiate a `java.security.SecureRandom` object. Do not use `getInstance()`, `getInstanceStrong()` and `java.util.Random`.

File: pom.xml [hunk 8]
```diff
@@ -37,7 +37,8 @@
         <module>powertools-validation</module>
         <module>powertools-test-suite</module>
         <module>powertools-cloudformation</module>
-        <module>powertools-idempotency</module>
+        <module>powertools-idempotency</module> 
+        <module>examples</module>
```
[the remark this answers is not in the document]
Comment (kozub) [reply pom.xml:41]: What's the build time? Does it have any side effects? 
If there are no issues with mentioned above questions, I believe that it's good to have it built as a part of build process to have quick feedback when some changes are done in the project (version of PowerTools used in examples might be an issue here...)

[the remark this answers is not in the document]
Comment (scottgerring) [reply pom.xml:41]: Build time: +approx 45s, [3:30](https://github.com/awslabs/aws-lambda-powertools-java/actions/runs/4185461706/jobs/7252546650) up from [2:44](https://github.com/awslabs/aws-lambda-powertools-java/actions/runs/4133563939/jobs/7143632548)

The only other side-effect is that changes to the powertools that break the examples will have to fix the examples, but I'd say that is desirable!

File: pom.xml [hunk 9]
```diff
@@ -450,7 +452,7 @@
         <profile>
             <id>build-with-spotbugs</id>
             <activation>
-                <activeByDefault>true</activeByDefault>
+                <activeByDefault>false</activeByDefault>
```
Comment (scottgerring) [inline pom.xml:455]: I've turned this off by default, as it 1/ helps with the SAM tooling for the examples and 2/ speeds up local builds that haven't opted into it.

The github workflows explicitly activate and de-activate these using the profiles, so it has no impact on CI.

**Comment by scottgerring** [comment]
## Changeset Structure

The changeset appears large, but the majority of it is simply pulling in the code from the samples. I have included it all at once because 1/ we can have an clean cutover between the two repositories and 2/ breaking up e.g. by sample will make it harder to review accurately - all of the samples should have had the same two light modifications.

This changeset is intended to modify things the minimum amount required incorporate the samples back into this branch. This minimizes the amount of actual change that needs to be reviewed in this PR. I expect subsequent work to revamp and ill out the sample set to happen in subsequent PRs. 

### 1. lambda-powertools - existing files
* build.yaml - workflow has been modified to trigger on changes in samples
* pom.xml - build has been modified to 1/ include samples sub-module and 2/ disable by-default spotbugs run. This is enabled and disabled explicitly by the build workflows anyway, and easy to opt into locally as needed with the associated profile
* spotbugs-exclude.xml - add an exclude for a bug in the samples. The samples did not previously have spotbugs run against them

### 2. Samples
Each sample has been lightly modified to fit into the powertools build:

* POM modified to 1/ add samples parent, 2/ remove compiler version (supplied by surrounding build) , 3/ modifiy lambda-powertools depedencies to use the current build version
* `CodeUri` set to `.` to use the maven build for each function
* In cases where `mockito` was in use, rolled back to the last Java-8 compatible stable version 4.1.1, so that the samples can run on Java 8 builds like everything else too
* `Function` subdirectory structured flattened out to fit maven hierarchy

File: pom.xml (hunk 8 above)
[the remark this answers is not in the document]
Comment (msailes) [reply pom.xml:41]: Yes, that would be desirable, we'd have a mechanism to keep the samples up to date.

File: .github/workflows/build.yml [hunk 10]
```diff
@@ -16,11 +16,14 @@ on:
...
       - 'powertools-metrics/**'
       - 'powertools-test-suite/**'
+      - 'powertools-examples/**'
       - 'pom.xml'
+      - 'powertools-examples/pom.xml'
       - '.github/workflows/**'
   push:
     branches:
       - master
+      - import-examples
```
Comment (msailes) [inline .github/workflows/build.yml:26]: Will this be removed in a later PR?

File: README.md (hunk 1 above)
Comment (msailes) [reply README.md:83]: See the examples directory **for** example **projects** showcasing usage of different utilities.

File: examples/README.md [hunk 11]
```diff
@@ -0,0 +1,4 @@
+## aws-lambda-powertools-examples
+
+This directory holds example projects demoing different components of the java powertools.
```
Comment (msailes) [inline examples/README.md:3]: replace java powertools with "Lambda Powertools for Java" or "Lambda Powertools"

**Comment by msailes** [comment]
Thanks for the really descriptive comment on the PR, that really helps.

**Review by msailes — APPROVED** [review]
LGTM

File: .github/workflows/build.yml [hunk 12]
```diff
@@ -16,7 +16,9 @@ on:
       - 'powertools-parameters/**'
       - 'powertools-metrics/**'
       - 'powertools-test-suite/**'
+      - 'powertools-examples/**'
```
Comment (jeromevdl) [inline .github/workflows/build.yml:19]: isn't it `examples` only ?

File: .github/workflows/build.yml [hunk 13]
```diff
@@ -16,7 +16,9 @@ on:
...
       - 'pom.xml'
+      - 'powertools-examples/pom.xml'
```
Comment (jeromevdl) [inline .github/workflows/build.yml:21]: isn't it `examples` only ?

File: .github/workflows/build.yml [hunk 14]
```diff
@@ -33,7 +35,9 @@ on:
...
```
Comment (jeromevdl) [inline .github/workflows/build.yml:40]: isn't it `examples` only ?

File: pom.xml (hunk 8 above)
[the remark this answers is not in the document]
Comment (jeromevdl) [reply pom.xml:41]: +1

File: pom.xml [hunk 15]
```diff
@@ -77,6 +78,7 @@
         <junit-jupiter.version>5.9.2</junit-jupiter.version>
         <aws-embedded-metrics.version>1.0.6</aws-embedded-metrics.version>
         <jmespath.version>0.5.1</jmespath.version>
+        <projectversion>1.13.0</projectversion>
```
Comment (jeromevdl) [inline pom.xml:81]: why this variable ?

**Review by jeromevdl — COMMENTED** [review]
Should the example poms have the examples/pom.xml as parent or the lambda-powertools root pom?

(See my other comments)

**Comment by scottgerring** [comment]
> Should the example poms have the examples/pom.xml as parent or the lambda-powertools root pom?
> 
> (See my other comments)

yes - silly mistake - some were, some weren't. fixed.

**Comment by kozub** [comment]
Looks good to me. 

One minor suggestion: 

I think that we might names of 2 examples packages in theirs pom.xml to have it more consistent: 

Right now when we run all tests it produces logs (please ignore info about failing test): 
```
[INFO] Reactor Summary for AWS Lambda Powertools for Java library Parent 1.14.0:
[INFO]
[INFO] AWS Lambda Powertools for Java library Parent ...... SUCCESS [  0.311 s]
[INFO] AWS Lambda Powertools for Java library Core ........ SUCCESS [  2.522 s]
[INFO] AWS Lambda Powertools for Java library Serialization Utilities SUCCESS [  2.596 s]
[INFO] AWS Lambda Powertools for Java library Logging ..... SUCCESS [  3.538 s]
[INFO] AWS Lambda Powertools for Java library Tracing ..... SUCCESS [  2.695 s]
[INFO] AWS Lambda Powertools for Java library SQS ......... SUCCESS [  2.864 s]
[INFO] AWS Lambda Powertools for Java library Metrics ..... SUCCESS [  2.440 s]
[INFO] AWS Lambda Powertools for Java library Parameters .. SUCCESS [  3.165 s]
[INFO] AWS Lambda Powertools for Java validation library .. SUCCESS [  3.167 s]
[INFO] AWS Lambda Powertools for Java library Test Suite .. SUCCESS [  2.547 s]
[INFO] AWS Lambda Powertools for Java library Cloudformation SUCCESS [  2.086 s]
[INFO] AWS Lambda Powertools for Java library Idempotency . SUCCESS [  4.064 s]
[INFO] AWS Lambda Powertools for Java library Examples .... SUCCESS [  0.004 s]
[INFO] A sample Hello World created for SAM CLI using powertools core utilities SUCCESS [  3.381 s]
[INFO] A sample Hello World using powertools idempotency .. FAILURE [  8.181 s]
[INFO] Sample app demoing parameters utility of Powertools. SKIPPED
[INFO] Sample app demoing serialization utility of Powertools. SKIPPED
[INFO] Sample app demoing SQS Batch Processing utility of Powertools. SKIPPED
[INFO] Sample app demoing validation utility of Powertools. SKIPPED
```

**Comment by scottgerring** [comment]
> Looks good to me.
> 
> One minor suggestion:
> 
> I think that we might change names of 2 examples packages in theirs pom.xml to have it more consistent.
> 
> Right now when we run all tests it produces logs (please ignore info about failing test):
> 
> ```
> [INFO] Reactor Summary for AWS Lambda Powertools for Java library Parent 1.14.0:
> [INFO]
> [INFO] AWS Lambda Powertools for Java library Parent ...... SUCCESS [  0.311 s]
> [INFO] AWS Lambda Powertools for Java library Core ........ SUCCESS [  2.522 s]
> [INFO] AWS Lambda Powertools for Java library Serialization Utilities SUCCESS [  2.596 s]
> [INFO] AWS Lambda Powertools for Java library Logging ..... SUCCESS [  3.538 s]
> [INFO] AWS Lambda Powertools for Java library Tracing ..... SUCCESS [  2.695 s]
> [INFO] AWS Lambda Powertools for Java library SQS ......... SUCCESS [  2.864 s]
> [INFO] AWS Lambda Powertools for Java library Metrics ..... SUCCESS [  2.440 s]
> [INFO] AWS Lambda Powertools for Java library Parameters .. SUCCESS [  3.165 s]
> [INFO] AWS Lambda Powertools for Java validation library .. SUCCESS [  3.167 s]
> [INFO] AWS Lambda Powertools for Java library Test Suite .. SUCCESS [  2.547 s]
> [INFO] AWS Lambda Powertools for Java library Cloudformation SUCCESS [  2.086 s]
> [INFO] AWS Lambda Powertools for Java library Idempotency . SUCCESS [  4.064 s]
> [INFO] AWS Lambda Powertools for Java library Examples .... SUCCESS [  0.004 s]
> [INFO] A sample Hello World created for SAM CLI using powertools core utilities SUCCESS [  3.381 s]
> [INFO] A sample Hello World using powertools idempotency .. FAILURE [  8.181 s]
> [INFO] Sample app demoing parameters utility of Powertools. SKIPPED
> [INFO] Sample app demoing serialization utility of Powertools. SKIPPED
> [INFO] Sample app demoing SQS Batch Processing utility of Powertools. SKIPPED
> [INFO] Sample app demoing validation utility of Powertools. SKIPPED
> ```

I've changed the example project names to match the general format, and i've put a publish exclusion on them so they don't end up in maven central.

**Review by jeromevdl — APPROVED** [review]

**Comment by scottgerring** [comment]
Created change in aws-samples/aws-lambda-powertools-examples/pull/296 to forward viewers onto the samples here.

**Review by jeromevdl — APPROVED** [review]
Good for me

**Pull request merged**